include/hydrogen/device/gpu/cuda/CUDALaunchKernel.hpp

```diff
@@ -8,24 +8,90 @@
 
 #include "CUDAError.hpp"
 
+#include <array>
+#include <tuple>
+#include <utility>
+
 namespace hydrogen
 {
 namespace gpu
 {
+namespace details
+{
+
+template <typename... Ts, size_t... Is>
+auto pack_arguments_impl(std::tuple<Ts...>& args,
+                         std::index_sequence<Is...>)
+{
+    return std::array<void*, sizeof...(Ts)>{{&std::get<Is>(args)...}};
```

Comment:
As we discussed, I'd expect we'd need to handle const pointer args:
```suggestion
    return std::array<void*, sizeof...(Ts)>{{const_cast<void*>(reinterpret_cast<const void*>(&std::get<Is>(args)))...}};
```
But it seems to work as-is, so it might not be necessary.